### ISSUE TITLE
Nested virtualisation support

### ISSUE THREAD

**Issue by mickys** [google/android-emulator-hypervisor-driver-for-amd-processors#20]
Hi, since this is based on KVM ( which has nested virtualisation implemented ), it would pretty much allow a Windows 10 Guest running on a KVM host to run the emulator.

I'm not sure how much work this would involve. But i sure could use it.

If this is not on the table i would love to get some pointers as to what needs to get implemented for this to work, maybe me or someone else gives it a shot.

Thanks.

**Comment by Taogle2018** [comment]
Running nested virtualization (my understanding) is typically the responsibility of L1 hypervisor (KVM in your case). A L2 hypervisor should work if it can already run well on bare metal. 
I did run a few tests in nested. However, I did not get any work for AMD. Tested with KVM, VirtualBox and VMWare workstation.
For Intel, it is another story. I can run nested with KVM and Hyper-V. With VMWare Workstation, I can run the emulator but it got stuck at boot animation. You can set env variable GVM_ENABLE_INTEL=1 if you would like to try this on Intel.
I do not have any data but I tend to believe nested support in KVM for AMD should be good but may have a few bugs. But there is no much work here.

**Comment by mickys** [comment]
Pardon my ignorance as i mostly work with orchestrators ( Libvirt ) and kernel drivers for AMD SEV but would love to learn more and get involved higher up in the chain.

So in my case where KVM is L1 and GVM is L2
The emulator ( qemu ) reports
```
host doesn't support requested feature: CPUID.80000001H:ECX.svm [bit 2]
```
To me this suggests GVM would need handle this by talking to the L1 ?

I do understand there's a big difference between a KVM linux guest where the translation happens inside the kernel in comparison to Fully virtualised guest like windows.

----- 
Unrelated to this one could maybe run the android emulator avd directly on the KVM Host and connect adb / vnc into the output from the guest ? ( still would need a GPU for acceleration on the host )

**Comment by mickys** [comment]
Thanks for explaining :)

In the end i got a lineageos VM setup on the host, with VNC and using adb remote with reverse. Seems to do the job just fine for what my students are doing. 

I'll test out the swiftshader ( on guest ) once i manage to get the qemu command converted to a proper libvirt domain xml. Do i need to use a special client for the acceleration to work client side ? i wouldn't expect something like Spice or VNC to support it ( maybe some kind of web client ? ).

I would definitely love to run the Google provided AVD images thou for that i suppose i need to setup the android emulator on the host then create a custom domain xml that passes the correct parameters to the "android qemu emulator" binary, then figure out the video client with acceleration support.

Thank you again for your time :)

**Comment by Taogle2018** [comment]
I am sorry I cannot answer this question because I am not an export in this area.
https://github.com/google/gpu-emulation-stress-test
I would suggest that you contact the developer here. He knows much more.

**Comment by lucaszanella** [comment]
I'm trying to run the android emulator on ubuntu 20.04 guest on ubuntu 20.04 host. When I launch, it warns me that it will be slow because of nested virtualization, but it's much slower than I predicted. It takes 10 minutes to boot a x86_64 image. Is something wrong?

I tried this on my i7 11th gen notebook and as well as on my Ryzen 7 2700x with 16 cores to the VM, and in both it's much slower than it should be. 

I thought this had to do with graphics but on the Ryzen I did a GPU passtrhough and still things were slow. On the i7 I activated OpenGL virtualization acceleration through virtio and still very very slow.

How to run nested virtualization fast?
